### PR TITLE
[JavaScript] Add new concept excercise: array-destructuring

### PR DESCRIPTION
**Pull request by kristinaborn**
Issue: #1893 

This is ready for review. Tagging @SleeplessByte directly since I'm not in the Slack channel.

### CODE REVIEW

File: languages/javascript/exercises/concept/array-destructuring/.docs/after.md
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-destructuring/.docs/after.md:1]: Maybe we can link back to MDN on `destructuring assignment`: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment, same for `array literal expression`.

File: languages/javascript/exercises/concept/array-destructuring/.docs/after.md
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-destructuring/.docs/after.md:5]: This actually doesn't compile! 

```suggestion
const [french80, french90, french100] = frenchNumbers
```

File: languages/javascript/exercises/concept/array-destructuring/.docs/after.md
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-destructuring/.docs/after.md:15]:
```suggestion
Because variables are mapped to values in the array by position, destructuring syntax can be used to assign or re-assign multiple variables in a single expression.
```

File: languages/javascript/exercises/concept/array-destructuring/.docs/after.md
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-destructuring/.docs/after.md:22]: This example shows nesting. I think it's better to first show a swap:

```javascript
let [a, b] = ['world', 'hello']
;[b, a] = [a, b]

a
// => 'hello'

b
// => 'world'
```

And then have a section on nesting.

File: languages/javascript/exercises/concept/array-destructuring/.docs/after.md
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-destructuring/.docs/after.md:32]:
```suggestion
The syntax allows to skip values when mapping, for example to ignore specific positions in the array.
```

File: languages/javascript/exercises/concept/array-destructuring/.docs/after.md
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-destructuring/.docs/after.md:36]:
```suggestion
getUserInfo()
// => ["Valerie", "Noir", "Sunny Lane 523"]

const [, , streetAddress] = getUserInfo()

streetAddress
// => "Sunny Lane 523"
```

File: languages/javascript/exercises/concept/array-destructuring/.docs/after.md
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-destructuring/.docs/after.md:39]:
````suggestion
The assignment is also not required to use all the values.

```javascript
const [firstName, lastName] = getUserInfo()

firstName
// => "Valerie"

lastName
// => "Noir"
```
````

File: languages/javascript/exercises/concept/array-destructuring/.docs/after.md
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-destructuring/.docs/after.md:41]:
```suggestion
It's even possible to extract _more_ values than the array contains; the leftover variables will be assigned `undefined`. This may be useful when the amount of values isn't known ahead of time.
```

File: languages/javascript/exercises/concept/array-destructuring/.docs/after.md
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-destructuring/.docs/after.md:44]:
```suggestion
const pickAtLeastOne = ['first choice', 'second choice']
const [first, second, third, fourth] = pickAtLeastOne

first
// => "first choice"

second
// => "second choice"

third
// => undefined

fourth
// => undefined
```

File: languages/javascript/exercises/concept/array-destructuring/.docs/after.md
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-destructuring/.docs/after.md:53]:
````suggestion
```

The array destructuring assignment can provide _default values_ in case there is none in the source array.

```
const [first, second, third="none selected", fourth] = pickAtLeastOne

third
// => 'none selected'

fourth
// => undefined
```
````

File: languages/javascript/exercises/concept/array-destructuring/.docs/after.md
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-destructuring/.docs/after.md:55]:
```suggestion
Javascript has a built-in operator that makes it easier to work with indefinite numbers of elements. When `...` appears on the left-hand side of an assignment, those three dots are known as the `rest` operator. The three dots together with a variable name is called a rest element. It collects zero or more values, and stores them into a single array.
```

Comment (SleeplessByte) [reply languages/javascript/exercises/concept/array-destructuring/.docs/after.md:55]: Maybe also link back to MDN here.

File: languages/javascript/exercises/concept/array-destructuring/.docs/after.md
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-destructuring/.docs/after.md:68]:
```suggestion
Note that in JavaScript, unlike some other languages, a `rest` element cannot have a trailing comma. It _must_ be the last element in a destructuring assignment. The example below throws a `SyntaxError`:
```

File: languages/javascript/exercises/concept/array-destructuring/.docs/after.md
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-destructuring/.docs/after.md:74]:
```suggestion
When `...` appears on the right-hand side of an assignment, it's known as the `spread` operator. It expands an array into a list of elements. Unlike the rest element, it can appear anywhere in an array literal expression, and there can be more than one.
```

File: languages/javascript/exercises/concept/array-destructuring/.docs/after.md
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-destructuring/.docs/after.md:82]:
````suggestion
const woow = ['A', ...oneToFive, 'B', 'C', 'D', 'E', ...oneToFive, 42]

woow 
// =>  ["A", 1, 2, 3, 4, 5, "B", "C", "D", "E", 1, 2, 3, 4, 5, 42]
```
````

File: languages/javascript/exercises/concept/array-destructuring/.docs/instructions.md
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-destructuring/.docs/instructions.md:7]:
```suggestion
Elyse will summon the first card in the deck, without using the `array[index]` or `.shift()`. It's just like magic.
```

File: languages/javascript/exercises/concept/array-destructuring/.docs/instructions.md
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-destructuring/.docs/instructions.md:18]:
```suggestion
Elyse performs sleight of hand, and summons the second card in the deck, without using the `array[index]`.
```

File: languages/javascript/exercises/concept/array-destructuring/.docs/instructions.md
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-destructuring/.docs/instructions.md:29]:
```suggestion
Elyse will make the top two cards of the deck switch places. She doesn't need to call a single function.
```

File: languages/javascript/exercises/concept/array-destructuring/.docs/introduction.md
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-destructuring/.docs/introduction.md:3]:
```suggestion
In this example, each value in the `numberOfMoons` array is assigned to its corresponding planet:
```

File: languages/javascript/exercises/concept/array-destructuring/array-destructuring.js
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-destructuring/array-destructuring.js:18]: Can you extract this to a _top level constant_ instead? It can be right above this function.

File: languages/javascript/exercises/concept/array-destructuring/array-destructuring.js
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-destructuring/array-destructuring.js:1]:
```suggestion
@ts-check

export function getFirstCard(deck) {
```

([`@ts-check`](https://www.typescriptlang.org/docs/handbook/intro-to-js-ts.html#ts-check))

File: languages/javascript/exercises/concept/array-destructuring/array-destructuring.js
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-destructuring/array-destructuring.js:5]: It's not a requirement, but if you're up for it, can we add a bit of JSDoc? The reason for this is that it will allow v3 in-browser coding and vscode/jetbrains IDE people to see type information! You don't *need* to do this. I'll create an improvement issue if you don't.

[Here is an example from another exercise.](https://github.com/exercism/v3/blob/master/languages/javascript/exercises/concept/array-analysis/array-analysis.js#L3-L10)

```javascript
/**
 * Get the position (index) of the card in the given stack
 *
 * @param {number[]} stack
 * @param {number} card
 *
 * @returns {number} position of the card in the stack
 */
export function getCardPosition(stack, card) {
  throw new Error('Implement the getCardPosition function')
}
```

File: languages/javascript/exercises/concept/array-destructuring/array-destructuring.spec.js
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-destructuring/array-destructuring.spec.js:16]: Ha. I love how this uses the concept without _really_ giving it away.

File: languages/javascript/exercises/concept/array-destructuring/array-destructuring.spec.js
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-destructuring/array-destructuring.spec.js:43]: Do the simple one first.

File: languages/javascript/exercises/concept/array-destructuring/array-destructuring.spec.js
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-destructuring/array-destructuring.spec.js:26]: Do the simple one first. Maybe also add an array with _no_ items, and array with _1_ item.

File: languages/javascript/exercises/concept/array-destructuring/array-destructuring.spec.js
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-destructuring/array-destructuring.spec.js:13]: Do the simple one first

File: languages/javascript/exercises/concept/array-destructuring/array-destructuring.spec.js
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-destructuring/array-destructuring.spec.js:61]: Simple one first

File: languages/javascript/exercises/concept/array-destructuring/example.js
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-destructuring/example.js:17]:
```javascript
export function swapTopTwoCards([a, b, ...rest]) {
  return [b, a, ...rest]
}
```

File: languages/javascript/exercises/concept/array-destructuring/array-destructuring.spec.js
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-destructuring/array-destructuring.spec.js:78]: Add a test for _empty_ array!

File: languages/javascript/exercises/concept/array-destructuring/example.js
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-destructuring/example.js:30]: Should use the top-level constant (as mentioned)

```suggestion
export function insertFaceCards([head, ...tail]) {
  return [head, ...FACE_CARDS, ...tail]
}
```

**Review by SleeplessByte — COMMENTED** [review]
I _really_ like what you did @kristinaborn . See in-line comments. 

@iHiD this is a typical case where we can introduce a custom array class that blocks function calls and blocks indexer, or... use an analyzer. I prefer the latter. 

@ErikSchierboom after this is merged, can you run your tool to explore `after.md` into the correct files? I don't feel like it's good time spent to make Kristina split it out right now.

**Comment by SleeplessByte** [comment]
@exercism/github-actions can you look at this PR's actions? Are the draft results different than non-draft results? I would expect to see all green and not 5 green, 4 pending.

**Comment by SaschaMann** [comment]
> can you look at this PR's actions? Are the draft results different than non-draft results? I would expect to see all green and not 5 green, 4 pending.

They're not, you need to rebase because the names of the required checks have been changed lately in #2541 (/rebase should work)

**Comment by SleeplessByte** [comment]
@kristinaborn when you process these comments, please also rebase ⚔️

File: languages/javascript/exercises/concept/array-destructuring/.docs/after.md
Comment (kristinaborn) [reply languages/javascript/exercises/concept/array-destructuring/.docs/after.md]: Yikes, I really should've caught this one. Changed.

File: languages/javascript/exercises/concept/array-destructuring/array-destructuring.spec.js
Comment (kristinaborn) [reply languages/javascript/exercises/concept/array-destructuring/array-destructuring.spec.js:12]: Done.

File: languages/javascript/exercises/concept/array-destructuring/array-destructuring.spec.js
Comment (kristinaborn) [reply languages/javascript/exercises/concept/array-destructuring/array-destructuring.spec.js]: Done.

File: languages/javascript/exercises/concept/array-destructuring/.docs/after.md
Comment (kristinaborn) [reply languages/javascript/exercises/concept/array-destructuring/.docs/after.md]: Done.

File: languages/javascript/exercises/concept/array-destructuring/.docs/after.md
Comment (kristinaborn) [reply languages/javascript/exercises/concept/array-destructuring/.docs/after.md]: Changed.

File: languages/javascript/exercises/concept/array-destructuring/.docs/after.md
Comment (kristinaborn) [reply languages/javascript/exercises/concept/array-destructuring/.docs/after.md]: Changed.

File: languages/javascript/exercises/concept/array-destructuring/array-destructuring.spec.js
Comment (kristinaborn) [reply languages/javascript/exercises/concept/array-destructuring/array-destructuring.spec.js:79]: Great idea! Added.

File: languages/javascript/exercises/concept/array-destructuring/example.js
Comment (kristinaborn) [reply languages/javascript/exercises/concept/array-destructuring/example.js]: Done.

File: languages/javascript/exercises/concept/array-destructuring/example.js
Comment (kristinaborn) [reply languages/javascript/exercises/concept/array-destructuring/example.js]: Done.

File: languages/javascript/exercises/concept/array-destructuring/.docs/after.md
Comment (kristinaborn) [reply languages/javascript/exercises/concept/array-destructuring/.docs/after.md:22]: Done.

**Comment by kristinaborn** [comment]
I've processed the comments and rebased :)

**Comment by SleeplessByte** [comment]
@kristinaborn this is looking great!

image

Did you perhaps miss this spoiler/expandable block? I had a few more. Most of them are just clarification/nit-picks; a lot with suggestions you can auto-apply.

**Comment by kristinaborn** [comment]
@SleeplessByte Yep, I completely missed those. Serves me right for working on a PR this late at night 😅 

I've applied the changes that could be done automatically, and I'll get to the rest tomorrow.

**Comment by SleeplessByte** [comment]
Sweet @kristinaborn . When you work on it, don't forget to run `/format` (prettier) on it. 

I'll merge it tonight, regardless, because the work is great!

**Comment by SleeplessByte** [comment]
(you can just ignore that, I ran prettier for you: https://github.com/exercism/v3/pull/2536/commits/7a78f7557192c63485dc3d0a2974365cf8b88e46 -- and when you push again, the checks will run again)

**Comment by ErikSchierboom** [comment]
@SleeplessByte Let me know when you want to merge. I can add a commit to it with the exploded files.

File: languages/javascript/exercises/concept/array-destructuring/array-destructuring.spec.js
Comment (kristinaborn) [reply languages/javascript/exercises/concept/array-destructuring/array-destructuring.spec.js:41]: Done.

File: languages/javascript/exercises/concept/array-destructuring/array-destructuring.spec.js
Comment (kristinaborn) [reply languages/javascript/exercises/concept/array-destructuring/array-destructuring.spec.js:25]: Done.

File: languages/javascript/exercises/concept/array-destructuring/array-destructuring.js
Comment (kristinaborn) [reply languages/javascript/exercises/concept/array-destructuring/array-destructuring.js:21]: I'm not very familiar with TypeScript but I took a crack at this anyway. Hopefully I didn't end up making more work for you instead of less 🤞

**Comment by kristinaborn** [comment]
@SleeplessByte I believe I've addressed all of your comments now. Please let me know if I've missed anything.

File: languages/javascript/exercises/concept/array-destructuring/array-destructuring.js
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-destructuring/array-destructuring.js:41]: This is perfect!

**Comment by SleeplessByte** [comment]
@ErikSchierboom this is ready when you are.

**Review by SleeplessByte — APPROVED** [review]

**Comment by SleeplessByte** [comment]
@kristinaborn your work here is done! Love it. Thank you. Erik will upgrade the file to the new format. I have yet to update our guide... so that. Let me know if you had fun doing this and hope to see you around -- you can low key join our v3 build/maintainers crew if you like.

**Comment by kristinaborn** [comment]
@SleeplessByte I had lots of fun! Thank you for all the excellent feedback. I'd love to low key join the crew and continue to help out where I can. v3 is going to be very cool.

**Comment by SleeplessByte** [comment]
This Saturday - October 31st, also known as Halloween - (and already "today" for many of you!) we're holding an Exercism v3 hackathon from 13:00 UTC to 19:00 UTC. It's the first opportunity to see the new staging site for the v3 website, get a feel for how your tracks are coming together, and work on moving them forward in preparation for launch.

We'll be the same place for video as normal (via the slack) and also in the #v3-markathon channel on Slack. Even if you can only drop by for 10 minutes to say hello, it would be lovely to see you! (Camera/webcam/video NOT mandatory)

If you can't make it, I'll send an update tomorrow with the details of where we've got to :)

To get slack access, on the site sign up as mentor.

---

As for maintainers, you can add yourself to the maintainers file in the JavaScript folder in this repository, then head over to the JavaScript repository and add yourself there too! 

The first one will make sure I know who I can tag  - the second will give you review access to this repository!

**Comment by SleeplessByte** [comment]
I'm just going to merge this and then make the other changes I did on a new PR because I can't push to this PR from _codespaces_.

Thank you for your patience @kristinaborn

**Comment by kristinaborn** [comment]
> To get slack access, on the site sign up as mentor.

@SleeplessByte I can't seem to figure out how/where to do this 🤔

**Comment by shubhsk88** [comment]
You can check the steps to join the slack group over here [here](https://exercism.io/mentoring-guide)

**Comment by SleeplessByte** [comment]
@kristinaborn let me know if you need additional help :)

**Pull request merged**
